Feature/imp end to end

cos opened the pull request:
**please review**

cos commented:
@bmmoore, @radumereuta, I think we should review and merge this. The kompile pipeline itself (`Kompile`) still needs refactoring, but we can leave that for after Dwight merges his configuration stuff also, and we can agree on an overall design.

@dwightguth, please review.

dwightguth commented on k-distribution/include/builtin/kast.k:
```diff
@@ -12,18 +11,20 @@ endmodule
   syntax KList ::= K
                  | ".::KList"           [klabel(#EmptyKList), hook(org.kframework.kore.EmptyKList)]
-                 | KList "," KList      [klabel(#KList), left, assoc, unit(#EmptyKList), hook(org.kframework.kore.KList)]
+                 | KList "," KList      [klabel(#dotKList), left, assoc, unit(#EmptyKList), hook(org.kframework.kore.KList), prefer]
```
pretty sure this klabel was right the first time...

dwightguth commented on k-distribution/include/builtin/misc.k:
```diff
@@ -0,0 +1,83 @@
+// Copyright (c) 2015 K Team. All Rights Reserved.
```
I am fine with this being included in the merge, but I just wanted to give you the heads up that given that this particular file is so utterly different from the K prelude we had before, I had to completely drop it in favor of my own implementation which is based more heavily on the original prelude, in order to run the C semantics. So this will eventually have to be deleted.

cos replied:
Yes, we'll have to agree on builtins. We may even discuss extending KORE with logical operators. At the very least, when merging your prelude, make sure to retain and use the LOGIC module from prelude.

dwightguth replied:
I will make sure to maintain the split Bool/KBool distinction you would like to see when I merge that part of my changes.

dwightguth commented on k-distribution/include/builtin/thekstuff.k:
```diff
@@ -0,0 +1,6 @@
+// Copyright (c) 2015 K Team. All Rights Reserved.
```
looks like this file isn't being used by anything, can we delete it?

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/AddParentCells.java:
```diff
@@ -0,0 +1,290 @@
+        }
+
+        // They were also not forced to be separate
+        throw new IllegalArgumentException("Ambiguous completion");
```
I see several errors in this file that look like they are things that could happen if the user writes their definition wrong. These errors need to use the KExceptionManager in order to print a friendly error message.

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/CloseCells.java:
```diff
@@ -0,0 +1,260 @@
+                if (required.isEmpty()) {
+                    return KApply(label, KList(contents));
+                } else {
+                    throw new IllegalArgumentException("Closed parent cell missing " +
```
same in this file

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/IncompleteCellUtils.java:
```diff
@@ -0,0 +1,80 @@
+        return isOpen(cell.klist().items().get(2));
+    }
+
+    private static void flattenCells(List<K> children, K item) {
```
I think we should consider importing Assoc.flatten from my automata2 branch of my fork, and using it in place of this particular method.

cos replied:
it seems like a simple enough recursive method -- we could refactor once your fork is merged. Added #1443 to track it.

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/SortCells.java:
```diff
@@ -0,0 +1,213 @@
+            if (parentCell == null) {
+                parentCell = cell;
+            } else if (!parentCell.equals(cell)) {
+                throw new IllegalArgumentException("Cell variable used under two cells, "
```
another source of IllegalArgumentException

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/SortCells.java:
```diff
@@ -0,0 +1,213 @@
+                    for (K item : k.klist().items()) {
+                        if (item instanceof KVariable) {
+                            if (var != null) {
+                                throw new UnsupportedOperationException(
```
not going to keep pointing these out, but it needs to be addressed throughout the entire diff

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/ConcretizationInfo.java:
```diff
@@ -0,0 +1,65 @@
+/**
+ * Created by brandon on 3/31/15.
+ */
+public class ConcretizationInfo {
```
since this class stores some important metadata, can I get javadoc on the methods in this class?

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/SortInfo.java:
```diff
@@ -0,0 +1,43 @@
+
+import java.util.Map;
+
+public class SortInfo {
```
javadoc

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/TransformKORE.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Created by brandon on 3/30/15.
+ */
+public class TransformKORE extends AbstractKORETransformer<K> {
```
javadoc

dwightguth commented on kernel/src/main/java/org/kframework/kore/compile/VisitKORE.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Created by brandon on 3/30/15.
+ */
+public class VisitKORE extends AbstractKORETransformer<Void> {
```
and again

dwightguth commented on kore/src/main/scala/org/kframework/compile/ConfigurationInfo.java:
```diff
@@ -0,0 +1,38 @@
+// Copyright (c) 2015 K Team. All Rights Reserved.
```
what exactly is a java file doing in src/main/scala?

dwightguth commented on kore/src/main/scala/org/kframework/compile/LabelInfo.java:
```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2015 K Team. All Rights Reserved.
```
same again

dwightguth commented on kore/src/main/scala/org/kframework/compile/LabelInfo.java:
```diff
@@ -0,0 +1,70 @@
+/**
+ * Basic label information for cell completion
+ */
+public class LabelInfo {
```
javadoc on these methods?

dwightguth commented:
Can you please remove the four commits by Radu on March 31 and April 1 from the commit history? they will eventually be removed anyway, since they broke the C semantics, and if I don't have to try to merge into them, my job of merging against this branch will go considerably smoother.

radumereuta commented on kernel/src/main/java/org/kframework/kore/convertors/KILtoInnerKORE.java:
```diff
@@ -224,7 +224,7 @@ public K apply(TermComment t) {
 
     private org.kframework.attributes.Att attributesFromLocation(Location location) {
-        Up up = new Up(KORE.self());
+        Up up = new Up(KORE.self(), Set("org.kframework.attributes"));
```
Is this supposed to be hard coded?

cos replied:
yes. we could move it to a constant, but yes.

cos commented:
I've changed the regular exceptions to KEM ones for most cases, but didn't link the location info yet -- we'll do it incrementally as needed. Also added a `criticalErrorForK` static method to `KExceptionManager`. Named it like this as it was generating an ambiguity with a `criticalError` taking `ASTNode` (some AST nodes in the backend are K now). We'll rename it when we get rid of ASTNode throughout.

cos commented:
so, everybody, is this ok for merge?

dwightguth commented:
lgtm. You should fix the merge conflicts though, whatever they are.

cos commented:
yes, I see the master moved a bit in the meantime... let's see...

cos commented:
in the meantime, @bmmoore, @radumereuta, are you ok with the merge?

radumereuta commented:
lgtm

bmmoore commented:
lgtm

ehildenb commented on k-distribution/include/kframework/builtin/ffi.md:
````diff
@@ -0,0 +1,177 @@
+
+```k
+                 | "#ffiCall" "(" Int "," List "," List "," List "," FFIType ")" [function, hook(FFI.call_variadic)]
+```
````
```suggestion
```k
  syntax Bytes ::= "#ffiCall" "(" Int "," List "," List "," List "," FFIType ")" [function, hook(FFI.call_variadic)]
```
```

ehildenb replied:
Otherwise it's confusing and you're not able to copy-paste this code block around to other locatiions.

ehildenb commented on k-distribution/include/kframework/builtin/ffi.md:
````diff
@@ -0,0 +1,177 @@
+
+```k
+                 | "#ffiCall" "(" Bool "," Int "," List "," List "," Int "," FFIType ")" [function]
+
````
```suggestion
  syntax Bytes ::= "#ffiCall" "(" Bool "," Int "," List "," List "," Int "," FFIType ")" [function]

```

ehildenb replied:
Another option would be to just declare all three together, with the documentation above all together.

ehildenb commented on kernel/src/main/java/org/kframework/parser/ParserUtils.java:
```diff
@@ -155,8 +155,15 @@ else if (di instanceof Require) {
+
                 ArrayList<File> allLookupDirectoris = new ArrayList<>(lookupDirectories);
-                allLookupDirectoris.add(0, currentDirectory);
+                allLookupDirectoris.add(1, currentDirectory);
```
1.  This looks like a typo in the name of this variable `allLookupDirectoris => allLookupDirectories`
2.  What is this change? The `0 => 1`.

radumereuta replied:
@dwightguth why was this changed from 0 to 1?
Someone on discord complained about kompile not finding their file: https://github.com/runtimeverification/k/issues/3695
